### ISSUE TITLE
Ability to check if an entity type's visibility restrictions apply to a given entity of that type

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#541]
In https://github.com/WordPoints/wordpoints/issues/536#issuecomment-253849466 we concluded that in order to properly implement the "points logs viewing restrictions API" proposed in that ticket we are going to need a means of checking whether visibility restrictions apply to an individual entity.

Currently we do have the restricted visibility interface that entities can implement, which will allow us to check if a particular user can view the entity. However, in the points logs restrictions API, we want to be able to tell whether a restriction applies to a particular entity (a particular post, for example). The logs from the hooks API need to have their visibility restricted based on the entities that were the args of the hook when it fired, so that is why the entities API is ultimately involved.

**Comment by JDGrimes** [comment]
This API is currently wrapped in the `wordpoints_entity_user_can_view()` function. One issue comes with the `'wordpoints_entity_user_can_view'` filter. Any code that hooks into that filter will not be able to provide information as to whether it applies to a particular entity or not (except when it concludes that the user is not allowed to view the entity, which is irrelevant in this use-case). So we may want to deprecate this filter, as we have done with similar filters for in the points logs API.

**Comment by JDGrimes** [comment]
The question, of course, is what would replace that filter, and how we'll actually implement this in general.

In general, I think that we can just introduce a new interface (or a "version 2" of the interface that we have now), and have it include a `viewing_restrictions_apply()` method, or something of the sort.

However, that will have a drawback, because some of the logic will have to be duplicated between those two methods, which will have to pretty much operate independently without storing any of the results in properties so that the other method can use them. (Or, it would be bad structure to encourage them to do this when there is no guarantee that the methods will be called in a particular order, unless we are adamant that they are private to the API and the API should only be accessed by the higher-level functions.)

But it also has the issue that it doesn't really work for any more-generic things that could previously have hooked into that filter. Currently the filter is unused, however, and perhaps we should question its legitimacy. Is there really anything that it could be used for? Mostly it seems likely that it would be used for individual entity types, even though I think we intentionally omitted such a filter preferring the entities themselves to be overridden in that case. I'm not sure that any other use-case for it really exists though. I mean, their isn't likely to exist any cross-entity APIs that could just hook in like that, unless of course they were built on top of our entity API itself. So there probably wouldn't be some plugin out there that we'd want to hook into this for in an integrating module or something, because it provided a cross-entity visibility restriction API. More likely that we'd provide a UI so that users could restrict various entities or something, and this filter would provide us a convenient way to hook in here. Although I'm not sure that that really makes sense, we can't control entity visibility throughout WordPress, so such a UI would have to relate to a particular context. And in that case, this generic filter would be less-helpful, and might not even be helpful at all.

So right now we have a filter that is unused, is going to cause problems, and may not really have  any valid use-cases to start with.

**Comment by JDGrimes** [comment]
However, I'm actually leaning toward not just doing a new interface, ad actually going for an API more similar to the one we're contemplating for the points logs, where we have a children class registry. In this case, the entity slugs would be the parents and the classes for each entity would be for each individual restriction for that type of entity.

Now, generally speaking we might not have multiple restrictions for each type of entity, and if we are really confident about that we could go with a regular class registry without the extra dimension. But the option to mess with entity restrictions without having to override the core entity classes seems like a win all around.

This also provides us the option to do as we plan to with the points logs, and allow some restriction rules to be registered for "all" entities. We don't even have to support that yet if we don't want to, but it would leave the option open (and really, I don't think there is any reason not to as it provides almost no overhead if it isn't being used).

Perhaps best of all, this allows us to implement a pattern similar to what we are doing with the points logs and construct the restriction object with the entity/entity ID so that the code that checks if the restriction applies to that particular entity can be more easily shared with the code that checks if the restriction should block the user.

**Comment by JDGrimes** [comment]
> I mean, their isn't likely to exist any cross-entity APIs that could just hook in like that, unless of course they were built on top of our entity API itself. So there probably wouldn't be some plugin out there that we'd want to hook into this for in an integrating module or something, because it provided a cross-entity visibility restriction API. More likely that we'd provide a UI so that users could restrict various entities or something, and this filter would provide us a convenient way to hook in here. Although I'm not sure that that really makes sense, we can't control entity visibility throughout WordPress, so such a UI would have to relate to a particular context. And in that case, this generic filter would be less-helpful, and might not even be helpful at all.

I'm starting to wonder whether this really applies to the whole API. Sure, it makes sense to restrict entity visibility, but doing that by registering restrictions here *and not hooking into WordPress core's APIs instead*, so that we'd end up with multiple rules per entity, doesn't really make sense to me at the moment. I fail to see the use-case. Yes, it might make sense to restrict a previously unrestricted entity, as there are no WordPress APIs for that. But why not hook directly into the `read_post` capability, for example? Unless you want to restrict viewing the info for posts in relation to a particular context, like points logs. In which case, use the points logs visibility restrictions API.

I guess that we can still harmlessly have a children class registry even if we only ever register a single rule per entity. And either way, I think that it makes sense to separate the restriction objects from the entity objects, for the reasons pointed out above in relation to logic shared between the methods.

**Comment by JDGrimes** [comment]
The reason I bring it up is because I'm trying to figure out what to name things. I'm trying to keep in mind the fact that in the future we may have other kinds of restrictions for entities, like who can edit them, etc. And also the pattern set by the interface, `WordPoints_Entity_Restricted_VisibilityI`. I was thinking `WordPoints_Entity_RestrictionI` for the interface, using `user_can()` as a method name instead of `user_can_view()`. And thinking  `WordPoints_Entity_Restriction_View_Post` as a pattern for the classes. But with that pattern it didn't make sense to have multiple rules per entity. And yet, any other pattern doesn't really make sense because the rule is generic to the entity itself. I mean, we could reference the caps API or something but it really isn't necessary. So I like that pattern, but then it won't make much sense when we come to register the class in a children registry.

I guess maybe we could go with `WordPoints_Entity_Restriction_View_Post_Read` as a pattern, but it is unnecessarily redundant.

**Comment by JDGrimes** [comment]
I wonder, if we went with a regular registry for now, if we could possibly change in the future? I suppose we would have to treat the new registry as a separate "app"/API. Though I guess we could provide back-compat handling, but we wouldn't be able to perfectly fulfill both interfaces, would we? It would maybe have to be more of a special case. But I guess it would work to do something like that without a second "app".

**Comment by JDGrimes** [comment]
Well, the problem wouldn't be with the registration functions, but the `get_all()` method would have a direct conflict in what code that was using it would expect it to return. That is something that we wouldn't really be able to resolve without implementing an alternative method. In other words, we'd end up with a confused situation.

So despite the uncertainty as to whether it is really needed, let's go with the child registry. I can't see that it really hurts anything. (And I guess maybe it would be easier to convert that to a regular registry in the future if we wanted to than vice versa.)

**Comment by JDGrimes** [comment]
One other thing though: what about if we introduce per-attribute/child viewing restrictions? If we have a children registry here, then what would we have for that? A grandchild registry? I think that probably we should actually go ahead and pursue this now, while we're at it.

But then, maybe we'd want that to work to arbitrary levels of depth anyway. Like a user can view the comments for a post, but just the authors, not the comment content. However, at the same time the user ~~could be allowed to view the comment content independently~~ I guess that doesn't make sense. What about, *wouldn't* be allowed to view the comment information independently at all. That particular example still doesn't make a whole lot of sense, I guess. Or maybe it does? Like, a user isn't allowed to see the comments on the post, but when they are looking at the post they are allowed to see who commented. So they are allowed to see the information that a particular user left a comment on that post, but not what they said.

That might make sense as a use-case, and maybe even not just in a particular context. But would it make sense withing our entities API? Because it would take a lot of work to implement within WordPress anyway, and so it would involve a lot of custom contexts, probably. Basically, custom UIs. So what this is partly about is making it simpler for WordPoints to work with custom WordPress content presentation schemes that are fairly customized, without having to hack the plugin to death. Which might make some sense, actually.

Main question is, do we want to pursue arbitrary depth right now, or just children. 

Such an API could I suppose be implemented through the "all" restrictions, if we support them. We didn't see the use before, but now I think I do. So maybe the best solution for the moment is to just do one restriction per entity API, but support generic restrictions so that a multi-restriction API could be introduced via that. And all the while, keep in mind, with naming and such, that this is an API that could expand in several ways in the future (like CUD restrictions).

Did we just pivot again? Oh yes we did.

**Comment by JDGrimes** [comment]
In that case, we could make the argument that we ought to just go for the new interface implemented by the entities/children themselves, instead of this API. But that brings up the issue of the filter again, as the only way to implement generic rules, which won't work, and also the issues with duplication between the two methods.

However, there is also another issue, and that is how we will store this information in a reasonably usable fashion. That is, we can descend way down into the entity hierarchy if we wanted to, but how would we carry this along with us? I guess without any concept of depth for the visibility restrictions, the question doesn't even make sense. We wouldn't have to descend within the registry, not until we introduce arbitrary depth, if that ever happens.

I was just prompted to think of this in considering how the entity child restrictions would be stored. I suppose they'll have to be in a separate registry, since our registries don't support both top-level and child classes together. Though of course, we could introduce something like that if we though that it was the best way forward here. I don't really think that it is necessary though. But maybe it would make for an easier transition to an arbitrary depth registry in the future.

**Comment by JDGrimes** [comment]
How do entity child restrictions affect the entity itself? I guess that they don't exactly. Really, the entity restrictions are about the entity itself, as a whole, being restricted. The user isn't just not allowed to view some aspect of it or any aspect of it, they aren't supposed to know that it exists. So maybe we should drop "view" and "visibility" from the basic entity restrictions API entirely, and reserve that only for the entity children.

On the other hand, perhaps a similar concept exists with entity children, or at least some types of entity children, like relationships, as well. A user might not be supposed to know whether an entity has any relationships (or maybe metadata/foreign attributes) of a particular type, not just not be allowed to view them. I guess that would only be if such attributes/relationships were dynamic, or rather, could exist for some entities of that type, and not others. ("Optional.")

The fact is, the entity restrictions API has always been influenced by the points logs visibility restrictions API. Which is the reason why its focus has always been visibility specifically. For the points logs, this is more true, since really I suppose we could display the other information for a log (date, number of points, user, etc.), but just not the log text. But maybe we do need to think about this a little bit more in regard to the logs as well, since right now folks could be using the API for the express purpose of restricting the log completely, not just the log text. Which is different, although right now it is all the same thing. But if we want the option to show a blank log entry in the future, we should open it up I think.

**Comment by JDGrimes** [comment]
Just to prove the point that restricted means that you shouldn't even be allowed to know whether it exists, note that attempting to view a private post will result in a 404, and [requests to change to a 403 were rejected](https://core.trac.wordpress.org/ticket/29829#comment:2). So we'er on the right track as far as that goes.

**Comment by JDGrimes** [comment]
## Registry/child registry/grandchild registry?

So as far as children registry vs not, we currently don't have the concept of optional attributes/relationships really, so right now we wouldn't have any restrictions of that type for entity children. And we wouldn't be mixing the visibility restrictions for entity children with the other ones.

Although this means that we'll really have two APIs here, an is restricted from user API, and a user can view API. I suppose for entities though that is really the same thing, if the user can't view the entity, they wouldn't be able to modify it in any way either. But they still might be allowed to know about its existence I guess. Though of course something restricted they wouldn't be allowed to do anything else with either. But then, how do you tell someone that an entity exists, without revealing any of its attributes/relationships? I suppose that within the hooks API, which is the only place that we use the entity API right now, there isn't anywhere that that makes any sense it all. It might make a little sense if you were listing entities, but then if you weren't listing all entities of that type you'd be revealing some information. But even if you were listing all entities, saying "another entity would be here" doesn't really make sense. So I don't think that we need an entity visibility API right now, just the restriction API.

So that means that right now we'd be storing the entities alone, without any children, unless we are trying to be forward-compatible with optional children. And we'll be putting the children separate from any entities, because we don't really have visibility restrictions for entities, just total restrictions.

**Comment by JDGrimes** [comment]
But that still doesn't answer how we should store the entity children, whether in a children class registry or grandchildren class registry, which we'd be creating. I guess we could even go with a regular registry, but use some kind of delimiting in the slugs, but that doesn't really make sense because we'll probably want to get all of the visibility restrictions that relate to the children of a particular entity at once, which in that case we couldn't do. So we do either need to use a children or grandchildren registry.

I guess the choice between the two again comes down to whether we want to support multiple restrictions per entity child. I suppose that if we do, we could go ahead and support arbitrary depth in case we decide to actually introduce that in the future. 

I guess as we've said, we probably wont' hurt anything by moving forward, so let's be as forward thinking as possible with this, and go for the arbitrary depth registry. Because when it comes to that, the idea of multiple children vs not only matters in how the registry is used, it doesn't have to be enforced at what level the classes can occur, that is just convention. But let's go ahead and allow multiple classes per entity and per entity child for both of these apps.

**Comment by JDGrimes** [comment]
One thing to note with a registry like this though, I don't think that the `get_all()` and `get_all_slugs()`, methods make sense. At least not for us right now. So I'm not going to bother with adding them to the interface right now. If we find a use for them we an add a new interface in the future.

What these do in the children class registry is return all of the children classes grouped together under their respective parents. So what they would really be like in a class of arbitrary depth might be a `get_grandchildren()` method vs a `get_all()` method. But we don't currently have a need for that either, as far as I can see.

**Comment by JDGrimes** [comment]
## Post types

Something else to consider, and which I hadn't thought of, is that we'll need to register the restriction dynamically for each post type. I guess that means we'll need to be able to grab a list of the post types that entities are registered for. That logic is currently internal to the entity registration function though. Let's abstract it out. And really we should probably do the same for the hooks, etc. But let's pursue all of that in a separate ticket.

**Comment by JDGrimes** [comment]
## Return restriction object

One issue here is that we have one function, but we'd know like to determine two different things: whether the entity is restricted from the user, and whether or not the restriction even applies. As such, I suppose that we really can't just return a single value, and perhaps it would be better to deprecate the function completely and instead allow the restriction object to be retrieved by whatever code wanted to know this info, and then they could call both the `user_can()` and `applies()` methods in whatever order/combination they want.

The main issue there though is that we may actually have multiple restrictions per entity, and from several different locations ("know" restrictions as I've elected to call the total restrictions, apply to every other type of restriction as well; "all" restrictions also have to be included). So what I was thinking was that maybe the best thing to do would return a wrapper restriction that would loop through all of these internally.

**Comment by JDGrimes** [comment]
The reason that we can't just perform the loop internally to the app itself via separate methods is that that will result in duplicate work, it makes more sense to keep a single instance of each class for performing both checks.

**Comment by JDGrimes** [comment]
## Entity restrictions and entity children

[Above](https://github.com/WordPoints/wordpoints/issues/541#issuecomment-253989077), we asked:

> How do entity child restrictions affect the entity itself?

However, now I think we need to ask how entity restrictions affect the entity children. I thought that I might have touched on that, but I guess maybe I didn't. Anyway, my gut feeling is that if an entity is restricted, all of its children should be also. But there is also this idea that maybe the entity shouldn't always affect the children. Like a user wouldn't be able to view private posts, but might be allowed to view some comments on them. Though that is really a different issue, since in that case the comments are related entities that have their own separate restriction rules. What we'd be talking about here is the ability to view the relationship itself, or entity attributes. But that doesn't really make sense when there is a total restriction on something, only a viewing restriction. That use-case really doesn't make sense, because the user wouldn't ever be able to know about the private post at all. So there'd be nowhere that they could go to see the comments. A password-protected post, yes, but then we're talking about just one attribute of that post being restricted, not the whole thing. However, I guess in theory you could have something where those comments were actually displayed in a custom context, and in that case you know that they relate to something, you just don't know what post. You know it exists though, by that, so then does it become just a viewing restriction? I guess that could partly depend on how it was presented, but I think that in general the concept doesn't really make sense.

So the idea that entity children might not be restricted when the entity is doesn't seem to make any sense for the "know" type restrictions. With view type restrictions, I guess maybe it might make sense in some circumstances.

But what I'm really trying to get to the bottom of is whether we should automatically check all of the parent restrictions for entity children internally, or whether we should leave that to the outside world. Really, if we do it internally, I think that that will give us the option of introducing parent overriding restrictions later on. But the concern I have is that if we do that we'll have the kind of issues that I was concerned about above, where code that is descending into the entity hierarchy will end up causing us to perform duplicate checks for the higher levels each time it descends lower. Though I guess really what we should do at that point is introduce a way of having this API descend with you, or optimize for that in some way.

I also just happened to think, if we support arbitrarily deep restrictions, how many levels should be ignored by an overriding child restriction? Or should that be variable and declared by the restriction itself? Anyway, that is more complexity than we want to pursue right now, especially when we can't really even think of a good use-case.

**Comment by JDGrimes** [comment]
So, doing this internally lets us handle it in one place and more easily expand the API in the future. Forcing it to be done on the outside requires us to remember to actually do it, but allows us to optimize for the particular situation when we do.

All things considered, if we need to use the API in this way, it would be best to provide a central core method of doing so, rather than having it come from the outside. So lets automatically check the parent entities for now, if we need to handle descending into hierarchies we'll find a way to support that.

**Comment by JDGrimes** [comment]
So now that we've decided to do this, maybe it is a good time to ask whether it is even possible. 😄 

The fact is, we can't ascend up a hierarchy that we don't even know about, from the entity child ID to the higher entity IDs. Actually, we don't even have the concept of entity child IDs anyway, so what we pass in is just the entity ID, and this is what we construct even the restrictions for entity children with (in our current mock-up, anyway). Maybe this should change, but at the moment I think it is fine. It means that yes, when checking entity children restrictions we can hop in and automatically check restrictions on the entity itself as well. However, we can't check the entity for an entity grandchild (the related entity in a relationship, for example), because we'll only have the grandchild (or (2n*(great))-grandchild) ID, not the ID of the entity itself. But that just means that we don't have to worry about our issues with descending hierarchies and all that. 

Another thought in that regard, is that we could provide a flag that would disable the parent entity check, like if it had already been performed or something. Though at this point I'm hesitant to even introduce that; a more full-fleshed solution at a later date would be superior, I think.

**Comment by JDGrimes** [comment]
## Constructing with user ID

Note that we've been constructing the restrictions with the user ID, but in https://github.com/WordPoints/wordpoints/issues/536#issuecomment-254649388 we decided that was a bad idea, so we'll be passing the user ID to `user_can()` instead.

>That way it will be impossible to accidentally have the user be taken into account in applies().

**Comment by JDGrimes** [comment]
In regard to entity children and checking the parent, at the moment we don't actually need to perform such a check in core, so I think I'll just leave that for later when we actually do perform child entity checks.

**Comment by JDGrimes** [comment]
## Back-compat

We still need to provide back-compat for the old per-entity restrictions and filters through an "all" restriction (unless there is some better way). Well, it is a must if we are completely deprecating the old function, `wordpoints_entity_user_can_view()`. I guess in theory we don't actually have to do that, although it would only be useful on occasions where we just want to know if a user can view an entity or not, and don't care that other users can't. However, the old filter stuff and the restricted entities will be expected to be taken into account even when folks are using the new API and not the function, so I guess we would still have to move this logic to a restriction so that it would apply there as well (or at least that it the cleanest way to have the logic still apply to the new API, I think).

**Comment by JDGrimes** [comment]
This means that we're also deprecating an interface, which is a first. I guess we should give some kind of deprecated notice for it inside of the legacy restriction checker. That will only be thrown when the check for that entity is performed, not every time that the class is loaded, but I think that that is OK. It is doubtful that any custom code is implementing this interface yet anyway.

We will need to remember to also update the docs that relate to this interface though.

**Comment by JDGrimes** [comment]
Also, what should we do about the classes that currently implement that interface? I don't think that we need to force them to continue to implement it, but what should be done for code that will just assume that for the entity the `user_can_view()` method will exist, without checking for the interface? Technically that isn't a good idea, but perhaps we should still retain and deprecate the method? No, in this case, I don't think that that is necessary. However, in the future when deprecating interfaces we might find that we would need to retain the methods. So let's go ahead and set the pattern here anyway.

So we'll stop implementing the interface, but keep the methods and deprecate them.

**Comment by JDGrimes** [comment]
I'm not sure what function to use to give the deprecated warning for entities that implement the interface though. `_deprecated_argument()` doesn't really fit. I had thought that we could even go with `_doing_it_wrong()`, but that's not exactly what we want. We could go with `_deprecated_function()`, since it actually doesn't include any reference to function, even though that will seem weird. But perhaps that's our best option, as it indicates that something is deprecated. The issue is that we want to also indicate which entity/entity class it is that is triggering the error. I guess we can do that easily enough using `get_class()`, but `_deprecated_function()` doesn't give us the option to supply an extra message. So on second thought, `_deprecated_argument()` might actually work for us.

**Comment by JDGrimes** [comment]
## Getting the entity slug

This also reveals another issue, which is that up to now I've been just expecting the restrictions to be constructed with the entity ID. Though of course, I've been unsure of how and whether to work around the fact that our registries will automatically construct objects with their slugs in addition to any other args that are passed. However, I'm now realizing that for this legacy check we'll need not just the ID of the entity but also to know what kind of entity it is. That is, we need the entity object, or at least the entity slug so that we can retrieve the entity object. This wasn't an issue with single-entity restrictions, but it will likely be important for any generic restrictions like this one. I guess that if we let the restrictions be constructed with the slug and parent slug (as they would be anyway unless we make it possible for them not to be), then it would be possible to determine the entity slug. Of course, for entity children the entity ID would be the grandparent slug, and not just the parent slug (because the slug of the restriction itself would be registered as a child of the entity child, since there are multiple restrictions per entity/child), but all of the parent slugs would be passed as an array, so we'd be able to determine that.

So it seems like the best thing for us to do is to just allow the restrictions to be constructed with the slugs. However, there is also another issue here, because I have created a multi-restriction wrapper object that the API can return, instead of returning an array of restrictions. But the problem is that it implements the restriction interface, and we have also included the constructor there. However, it doesn't really make much sense to construct this wrapper with the slugs. At least, it will not have its own slug, though perhaps it isn't completely crazy to construct it with the entity slug. Though I guess we could just give it the slug `'wrapper'`. Or, we could just stop prescribing the constructor in the interface, since I'm not sure that it is the best practice anyway, although we've done it elsewhere.

**Comment by JDGrimes** [comment]
## Legacy restriction

Should the legacy restriction checker be registered as a "know" or a "view" restriction? Previously it was all about "view", but our two entities that had implemented the interface are now "know" restrictions. So I guess we should make it a "know" restriction, to be safe, in case anything else that is using it should also really be a "know" restriction. (Not that it really matters all that much at the moment since we are only checking for the view restrictions in core right now anyway, but still.)

**Comment by JDGrimes** [comment]
## Generic restrictions and entity children

I'm not sure if I addressed the idea of restrictions that apply to all children of an entity yet or not. I was just looking at how I had designed some of the logic in the mock up of the restrictions app and it looks like maybe that is what I was trying to do. However, I'm not sure that that really makes sense. So let's not implement it right now.

There is another question though, and that is how to handle the generic "all" restrictions, like our legacy restriction. I was thinking of them as applying only to the entities, that is, really only applying at the top level. But I hadn't thought about how that would mean we'd need to check if we were checking an entity child restriction before including the all restrictions in the check. But on thinking about it I guess that really that is the correct behavior, since we've decided not to pursue checking restrictions relating to the parents of entity children. So we will only have the "all" restrictions apply to the top level for now as well, I suppose.

**Comment by JDGrimes** [comment]
## First-level parent slugs

Another question that this has brought to my attention is what to do about top-level slugs. You see, our multilevel class registry accepts top-level parent slugs either as a simple string or as an array with just one element. However, it doesn't currently standardize this before constructing the classes, which means that every top-level class has to be able to deal with both situations (or if it doesn't this could cause issues). I guess probably the best thing to do here is to have the multilevel class registry always standardize these to be arrays. Even though strings are simpler to deal with in some ways, the fact that all lower levels will have arrays when the top level supports strings would be an odd inconsistency I think. And maybe we should go so far as to have the multilevel class registry only accept arrays for the parent slugs as well, just to be consistent and so devs know that their classes should always expect the parent slugs as an array. I do like the simplicity of being able to just pass a string when registering the classes, but I'm not sure that it is worth the potential uncertainty.

**Comment by JDGrimes** [comment]
I'm also wondering what pattern the generic restrictions' parent slugs should follow. I was just now going to have the parent slug of the legacy restriction be `array( 'all' )`, but I guess that we could also just use an empty parent slug for that, or rather have the classes be parentless, actually at the top level.

### `all`

```
- all
-- legacy
- post\post
-- status_nonpublic
```

Using `all` means that we are basically reserving that as an entity slug, so that no entity can have that slug. Which really isn't a problem, since that doesn't make sense as an entity slug anyway. There is the possibility that we'll want to do something similar to this with entity children in the future though, so we'd be potentially reserving that as an entity child slug as well. Still probably not a big problem though. Also this was the pattern that we intend to follow with the points logs restrictions, so it is nice for consistency.

### Top-level

```
legacy
- post\post
-- status_nonpublic
```

The benefit of registering the classes at the top level with no parent is that this might seem more intuitive in some ways. "This is registered a level above the entities, so naturally it applies to all of them." This could also be a pattern that we follow generically in regard to entity children later on, every higher level would be applied to the entity child (although we're not intending to pursue that right now and it isn't without some reservations). However, when we start talking about restrictions generic to all children of an entity this pattern breaks down, because it would mean that we'd use the entity slug itself to register the generic test that should be applied to all entity children, and we already register the entity restrictions as children of that. Which makes me think, isn't that what a generic restriction for all children of an entity is anyway? It is just a check on the entity itself, it doesn't take the child into account at all. Unless we're dealing with some kind of strange entity with dynamic relationships or something (dynamic because otherwise we'd be able to just register for each of them separately.) So yeah, I think that really restrictions generic to all entity children don't make any sense.

So in that case, following a pattern that could be used for entity children isn't a priority. In fact, it might be better for us to pursue top-level pattern because it naturally applies only to the top level, and restrictions that should occur across all entities.

I guess the only real downside that I can think of to this approach now is that it would be inconsistent with what we are doing with the points logs. However, that's only because we aren't using a multilevel registry for the points logs restrictions. We do have that option in theory though, if we introduce a multilevel children registry.

All things considered, I see no reason not to utilize the top level, seeing that we have it at our disposal, unless we have plans to use it for something else later. And as far as I can see, we don't.

**Comment by JDGrimes** [comment]
## Generic restrictions and entity slugs

But back to getting the entity slug, all of this is to say that generic restrictions will not be passed the entity slug, because it will not be a part of their slug, even if we construct the restrictions with their slug and parent slugs. So I guess we should pass in the actual slug hierarchy of the entity/child being checked as an extra parameter for the generic restrictions. Or maybe just pass it in to all restrictions, for consistency. Though in that case, it really isn't necessary to construct the class with the slugs at all. Currently we have no use for it, and because of the nature of the API I doubt we ever will have a need to get the slug of a particular restriction from the restriction object. But upon reflection I suppose that we should leave open the possibility that we would. Though if we're going to do that I guess we should include method on on the interface to get the slug and hierarchy. Though at present, we don't even expose the objects to the world, they are returned wrapped in a meta restriction object. We could allow the user to retrieve the list of restrictions, but they wouldn't even know what types they were. Unless we group the restrictions of each type together in wrappers and then let the wrapper we return wrap them. And for the greatest possible flexibility for ways of using the API in the future, maybe that is the way to go.

The potential use-case that I have in mind would be when something is wanting to get a list of restrictions that apply to a particular entity. Don't ask me why it would wan to do that though. Because the restriction slugs are completely arbitrary, and don't relate to any other APIs. Unlike the hook arg slugs, for example. Only if they would possibly relate to other APIs in the future would there really be a reason to do that. We have no reason at present to pull up a particular restriction or set of restrictions. So all things considered, we really have no need for this right now. And I think that we never will. The list of restrictions that apply to something shouldn't really be saved anyway, as it can change over time, and relying on a saved list could therefore lead to information disclosure.

*Ergo*, let's just make it so that we pass in the entity hierarchy as an extra arg to the restriction constructors, and make it so that the slug and parent slugs aren't passed in by the parent class. 

This will be a special option in the multilevel class. However, it would be nice not to have to pass this in each time, it would be better if we could construct the class with that setting so that it would then always know not to pass the slugs in. This makes sense because if the classes expect to be constructed without the slugs then they probably always expect to be constructed that way. So there is no reason to pass this option in each time we call for some classes. But though I've often thought of introducing such a feature, it is too much to pursue right now, because it would involve complex stuff with the apps as well as the registries. So we'll just pass in the args each time for now, it should all be internal to the app anyway.

**Comment by JDGrimes** [comment]
Or rather, I'll just add a method to the multilevel class to allow us to set this once and forget it. That way we can have the internal handling for that in the registry already be based on a property value instead of an arg passed into the methods. And we'll be able to just call this method in our class registry init function where we register our restrictions.

**Comment by JDGrimes** [comment]
Oops.

**Comment by JDGrimes** [comment]
## Unregistered entities

One more thing in regard to the legacy code, is that when we found that the entity didn't exist, we used to return false immediately, because we couldn't check if the user could view the entity when we didn't have the entity type object so that we could check for the interface. Even though we no longer have to rely on the interface now, I think that checking for the existence of the entity might still be a good idea. Because even though restrictions can exist without the entity, they will usually be registered by the same code that registers the entity. So if the entity is gone, it is likely that at least some of the restrictions relating to that entity will be gone as well. Which means it might not be safe to expose the entity to the user.

However, I don't think it is a good idea to leave this check as part of the legacy restriction, since that might cause it to accidentally get removed in the future.

And we might want to consider whether this is really a full solution. After all, some restrictions can be added by a module, and then that module can be deactivated, which means even though the entity might still exist it may not be safe to expose it. I think that that is more of an edge-case though, and not one that we are likely to encounter very often. So while we may want to handle this on a more per-restriction level at some point, I think that for now it is OK if we just check for the entity.

It isn't as simple as placing this check right in the method, because we always return a restrictions object not just a boolean value. So we either need to register an `all` restriction for this, leave it as part of the legacy restriction for now, or if we still want to hard-code this, provide a restriction object that we can preemptively instantiate and return that will return false in `user_can()`.

I think possibly the best thing to do is to add a generic restriction, because then we can deregister it in favor of a better solution in the future. Though I'm not sure that is really feasible, since we'd need to maintain this check indefinitely for back-compat, even after a better way was developed. Well, maybe not, perhaps we might decide to fall back to a particular capability or something in the future instead, or some people might want to do that. And having this separate would give them that opportunity to play with the logic, which is good as long as they know what they are doing. So let's go with a generic restriction for this, I guess.

**Comment by JDGrimes** [comment]
## Legacy filter

We've talked about how the legacy filter will no longer be applied on top of all of the other restrictions, but only inside of the legacy restriction. However, I just realized that it will also only end up being applied when the entity in question implements the legacy interface, and so the legacy restriction itself will apply. This is a bit of a conundrum, because we can't make the legacy restriction always apply or it will mess up our work in #424 and cause all points logs to appear as if they are restricted. It would be nice if we could check if the filter returned `false`, and then only mark the restriction as applicable in that case. But then it would need to be in the constructor, and unfortunately we can't move the it out of the `user_can()` method, since it expects the user ID to be passed. 

We could move it out of the restrictions altogether, but there is still a problem with that, because it expects the entity object to be passed, and although we can get that and only run the filter if it is available, we still would have to deal with figuring out what value to pass it for the `$can_view` parameter and how to deal with the return value if it changes it.

Considering that it is unlikely that anybody else is using this filter, I propose that we just don't worry about this and continue to apply it in the `user_can()` method of the legacy restriction, even though that might not be ideal.

**Comment by JDGrimes** [comment]
## GUIDs

I think we'll need to accept GUIDs here and switch to the correct context for the entity. However, that isn't possible when we are using the restriction objects, is it? Because they will be able to be used anywhere, that is, the `user_can()` method will not be called until after we've switched back to the correct context. So maybe we want to rethink that. I suppose that the wrapper could switch again when `user_can()` is called, but it would be nice not to have to switch twice (once when constructing the objects and again in `user_can()`). But I guess what we could do is have the wrapper go ahead and run the `user_can()` checks while we are still switched, save the result in a property, and then just return that later in `user_can()`. We already loop through all of the `applies()` methods in the wrapper constructor anyway. But it just makes me wonder whether we should just not return the wrapper at all then, and do something different instead. I guess the benefit of returning the wrapper is that we can decide to expand on that basic interface that it implements later to provide addition methods. Where as if we just return an array or something it will be less forward compatible.

So we'd still like to return an object like this, however, we can't just perform the `user_can()` checks in the constructor because we don't have the user ID there. ~~Though I guess we can actually go ahead and pass it in if we want to, since the wrapper already has the restrictions passed in specially.~~ No, we can't even do that, because we don't even have the user ID at that point, it doesn't get passed in to the restriction retriever method on our app, because it really isn't needed (except for here).

Which I guess means that we are left with switching twice or performing the switching externally to the app. Or, we could just demand that the user ID be passed in. But then we're back to the point of not returning a restriction object here, and just returning an array instead. Which really might be fine, I guess. It doesn't completely cramp us for the future, since we can always add more elements to the array. The only major downside is that it would not be optimized for checking multiple different users against the same set of restrictions. However, at present we don't really have a need for that. And I'm not sure that we ever would. Maybe if we were sending out messages to a list of users in relation to some entity, we'd need to check the list to make sure that only those users who could view the entity would get the message. So there would be potentially use-cases for that, but then we could just modify the app to accept an array of user IDs, or add another method, or something.

So I'm fine with moving forward on this with the user ID being forced to be passed in, and us just returning an array. We could still return a special object implementing the entity restriction interface, of course, but it would be odd because then we'd still have to pass the user ID in again to the `user_can()` method, so it wouldn't make much sense.

**Comment by JDGrimes** [comment]
However, that won't allow us to even check application without passing the user ID. And we currently are needing to do that in the points logs restriction that integrates with the entities API for the new Hooks API. So in the points logs restrictions' constructor we are checking the applicability of the entity restriction, even though we don't get the user ID passed in until `user_can_view()` is called. Adding a separate method or something that would allow us to check just applicability isn't a good option though, because we'll then have to repeat all of that stuff all over again in `user_can_view()`. But it seems like somewhere something is going to have to be duplicated, and if that's so, then I guess it might as well be the switching because a majority of the time there is no need to switch at all.

There is a fragile option here, and that is to just assume that the `user_can_view()` method of the points logs restrictions will always be called before the `applies()` method. That way we don't have to duplicate the logic, just place it in `user_can_view()` where we have the user ID, and set a flag on a property that `applies()` can read. Currently that assumption holds true, but I'm not sure that we can really consider it future safe.

Other option is to go back to expecting switching to be performed on the outside. Not ideal, and still wouldn't solve all of the problems because in the points logs restriction we'd still have to switch twice: once in the constructor, and again when `user_can_view()` was called. There's no moving it to the outside either, because we may be dealing with multiple args for a single points log, in an internal loop.

So it looks like we have little choice but to revert to the wrapper and double switching. The only alternative I can see would be a refactor of the points logs API so that each restriction was constructed with the user ID, but as we've discussed before, that is bad because it allows us to accidentally take the user ID into account when determining general applicability. But on the bright side of double-switching, we are already switching to the correct site for the log in question, so most of the time no switching will be needed for the args at all. The API as a whole still wouldn't be optimized for checking multiple users against the same set of restrictions though, if they related to an out-of-context entity. Which I think is actually quite acceptable since in most applications we should be able to work around that by going ahead and switching to the correct context to start with, as we do with the points logs.

**Comment by JDGrimes** [comment]
Now, when we are passed an entity GUID instead of just an ID, we need to be able to retrieve the entity ID from it so that we can pass just the plain ID to the entity restrictions when we construct them. This is pretty simple to do, we just have to know the entity slug, as that is the index of the entity ID in the GUID array. The question that comes up when getting the entity slug though is how to handle deep entity children. I was just going to do `$hierarchy[0]` to get the entity slug, but that would assume that that entity ID being passed was always for the entity at the top of the hierarchy. Which is true for entity children. But for entities embedded in the hierarchy, like related entities that would be grandchildren in the hierarchy, it is the deeper entity whose ID is expected.  I just sort of wanted to document that, because I'm not sure that we really said it explicitly above.

**Comment by JDGrimes** [comment]
## Slugless registry and tests

There is one unforeseen issue with our work to allow the multilevel class registry to be made slugless via a method call. We decided to call that method when the registry is initialized, which makes sense. However, it causes issues during the tests, because during the tests when we mock the apps the registry init actions don't get called. So the registries don't get set to be slugless. I think maybe the best thing to do for now, rather than setting this every time we mock the app during the tests, is to just create a subclass of the multilevel registry that is slugless by default. Then we can remove the methods to set this information for now and not have to worry about all of this or anything else that might come up. Not idea, but it is better than attempting to rework the whole apps API right now. So let's do that.

**Comment by JDGrimes** [comment]
## Empty hierarchy

I hadn't considered before what to do if we were passed an invalid hierarchy. We don't really validate entity slugs or entity child slugs, but I think there is one case that doesn't make sense, and that is when we are passed an empty hierarchy. The entity slug at least has to be provided, because we are going to be passed an entity ID after all. So it seems like if we are passed an empty hierarchy we should probably error in some way. Unless we feel that that is such an unlikely scenario that it would never happen. But I can foresee the possibility that a dev would accidentally pass an empty hierarchy without realizing it, and that this would be tough to debug if we don't return an error. However, if we just return `false` we'll kill something unless we force everybody to check that there wasn't an error. But if they are doing that, they should have just checked for the problem before the slug was passed in. But I guess it does leave us the possibility open to pass false out for other things in the future.

The alternative would be to just shortcircuit but return an empty wrapper instead. That wouldn't alert the dev to the problem though. But we could do that by calling `_doing_it_wrong()`.

We could also just do nothing. Which would result in the dev getting some kind of error because some of our generic restrictions expect the hierarchy to contain at least one element. I think maybe I'll just do that.

The only reason that this came up was because I had included an empty hierarchy in our tests, not fully thinking that that was invalid. And then when I realized that it was, I wondered whether we should have some more sanity checks. But I think it is OK if we just hope that the dev is sane in this case, since we're already expecting them to pass in values that match a lot of other criteria, and don't have any checks for that either.

**Comment by JDGrimes** [comment]
## Deprecate old function?

I'm not sure whether we actually need to deprecate `wordpoints_entity_user_can_view()`. I've sort of been planning to, but I've known that it might not be necessary. If you just need to check if a user can view an entity without needing to know about applicability, then it could be a nice shorthand function. Inside of loops and things though, it would be better to pull up the restrictions app just once instead. But I guess that is up to the developer to be on top of things. So let's keep this around for now, if we find a problem with it in the future, we can always deprecate then.

**Comment by JDGrimes** [comment]
## Generic restrictions and hierarchy

> I was just going to do $hierarchy[0] to get the entity slug, but that would assume that that entity ID being passed was always for the entity at the top of the hierarchy. Which is true for entity children. But for entities embedded in the hierarchy, like related entities that would be grandchildren in the hierarchy, it is the deeper entity whose ID is expected.

Forgot to do this with the unregistered restriction, I think.

But is that even really a problem? I mean, maybe we should really be checking all of the entities in the hierarchy. And as far as that goes, I guess we might as well go ahead and check entity children while we are at it, not just the entities.

I think I also forgot to properly use the hierarchy in the legacy restriction though. But I'm not sure what would be correct there. No, I think that in this case the legacy handling doesn't make sense for deeply embedded entities, only top-level ones, since it was all about top-level entities. So we should have the legacy restriction only apply when checking a top-level entity. Which it would anyway, wouldn't it? Yes, because of the way that it is registered, it is only registered for the top level.

But I guess that goes for the "unregistered" restriction as well, it is only registered for the top level, so that is the only thing that it will be used to check. So maybe we do need to introduce the ability to have restrictions that apply at all levels? We had kind of considered that before and we weren't sure that it made sense, but now I can see how it might for this. But I'm actually content to leave this as it is for now, since we aren't actually using deep entities, and maybe will never find a use for them. Though we should keep all of this in mind going forward.

**Comment by JDGrimes** [comment]
Closing notes:

- We deprecated the `wordpoints_entity_user_can_view` filter, and though we included legacy handling for it, it is imperfect. We decided that this is fine since the filter is probably not used anyway.
- We have retained the `wordpoints_entity_user_can_view()` function, as a convenience wrapper for the API.
- We deprecated the old interface for the entity objects, but provide legacy handling for it through a generic restriction.
- Classes that implemented the interface in core no longer do, but the `user_can_view()` method was retained on them, updated to use the new API, and deprecated.
- When entity classes are detected that still use the deprecated interface, a deprecated warning is given for them in the legacy restriction, using `_deprecated_arg()`.
- The new API consists of an app that has multiple deep multilevel registries (something we created as part of this effort) as children, each for different types of restrictions.
- There is currently a registry for "know" restrictions, and one for view restrictions.
- Each registry can have multiple restrictions registered for each entity/child, as well as generic top-level restrictions that apply to all entities.
- Restrictions can also apply to entity grandchildren to arbitrary depth, though this feature is currently unused in core.
- "know" restrictions can also apply to entity children, since it is possible for some entity children to be optional and for a user to not be allowed to know if a particular entity even has children of that type.
- For our deep registries, we decided not to implement `get_all()` or `get_all_slugs()` methods, or `get_grandchildren()` type methods, since we don't currently need them.
- The restrictions app itself is what is used to access the restrictions stored in the registries, they aren't accessed directly.
- They are accessed through a `get()` method on the app itself, which returns a custom restriction object that wraps any restrictions for the particular entity in question.
- Thus the restrictions themselves are never actually exposed individually. This is OK, since we've also decided not to pass them the slugs that they are registered with, because this information should not be saved as what restrictions relate to an entity is subject to change.
- The app's `get()` method automatically includes any "know" restrictions when other types of restrictions are being checked, and it also automatically includes the generic restrictions as appropriate (these only apply to the top-level entities, not the entity children).
- We currently don't automatically handle checking parent restrictions for entity children internally. When we have a use-case for that we should introduce a method for doing it accordingly. The reason for this omission is the complexity of it and the fact that we can't optimize for particular use-cases ahead of time. In the future however, we will probably add this kind of check internally, possibly with the option to have child entity restrictions override parent restrictions.
- Restrictions for an entity child are passed the entity ID only, just like restrictions for the entities themselves.
- We specifically decided not to construct the restrictions with the user ID, only the entity ID and slugs, since doing so would have made it possible to accidentally take the user into account when checking applicability, which should not be done as applicability only refers to the entity/child.
- The entity/child slug/s is always passed as an array to the restriction constructors, even when it contains only one element or the entity slug has been passed into `get()` as a string.
- The generic restrictions that apply to all top-level entities are registered at the top level of the registries, with an empty parent: `array()`.

**Comment by JDGrimes** [comment]
Developer docs are here: http://wordpoints.org/developer-guide/entity-api/restrictions/
